### PR TITLE
Added verification service

### PR DESCRIPTION
**Pull request by sbryzak**
Fixes https://issues.redhat.com/browse/CRT-774

### CODE REVIEW

File: pkg/verification/verification_service.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	SendVerification(signup *v1alpha1.UserSignup) error
+}
+
+// ServiceImpl represents the implementation of the verification service.
+type ServiceImpl struct {
+	config ServiceConfiguration
+}
+
+// NewVerificationService creates a service object for performing user verification
+func NewVerificationService(cfg ServiceConfiguration) (Service, error) {
```
Comment (alexeykazakov) [inline pkg/verification/verification_service.go:36]: you don't need to return `error`, do you?

File: pkg/verification/verification_service.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	toNumber := signup.Labels[v1alpha1.UserSignupPhoneNumberLabelKey]
+	client := twilio.NewClient(s.config.GetTwilioAccountSID(), s.config.GetTwilioAuthToken(), nil)
+	msg, err := client.Messages.SendMessage(s.config.GetTwilioFromNumber(), toNumber,
+		content, nil)
+	if err != nil {
+		return err
+	}
+
+	if msg.ErrorCode != 0 {
+		fmt.Printf(msg.ErrorMessage)
```
Comment (alexeykazakov) [inline pkg/verification/verification_service.go:63]: Let's return it's as an error, not print it.

File: pkg/verification/verification_service.go (hunk 1 above)
Comment (sbryzak) [reply pkg/verification/verification_service.go:36]: Good point, fixed.

File: pkg/verification/verification_service.go (hunk 2 above)
Comment (sbryzak) [reply pkg/verification/verification_service.go:63]: I think this block is unnecessary, I've removed it all together and added logging to the previous error handler.

**Comment by sbryzak** [comment]
@alexeykazakov This is actually my second PR, the first one I accidentally committed our Twilio auth token for one of our subaccounts which I was using in the tests.  I killed the first PR and deleted the subaccount immediately, thankfully it wasn't the token for the master account which would have been a more difficult problem to fix :)   Anyway about the tests, I was hoping to get some suggestions - there doesn't seem to be any kind of mock API available to use for Twilio, so I'm still trying to think of the best way to test the new verification service without actually having to invoke the Twilio client API.  Any ideas?

**Comment by alexeykazakov** [comment]
@sbryzak the twillo client uses http client behind the scenes, I guess? In that case we could use gock for example to mock the http client.
It's trickier with kube because go-client won't initiate without kube token present in the local file system, and this is why we don't have tests for our kube services yet, but it should not be a problem for testing twillo.

**Comment by sbryzak** [comment]
We can actually pass in an HttpClient to the Twilio client constructor (it's the third parameter):

`twilio.NewClient(s.config.GetTwilioAccountSID(), s.config.GetTwilioAuthToken(), nil)`

I'll see what I can do to allow us to inject a mock client here.

File: pkg/controller/signup_test.go [hunk 3]
```diff
@@ -217,3 +224,11 @@ func (m *FakeSignupService) GetSignup(userID string) (*signup.Signup, error) {
 func (m *FakeSignupService) CreateUserSignup(ctx *gin.Context) (*crtapi.UserSignup, error) {
 	return m.MockCreateUserSignup(ctx)
 }
+
+type FakeVerificationService struct {
+	MockSendVerification func(ctx *gin.Context, signup *crtapi.UserSignup) error
```
Comment (alexeykazakov) [inline pkg/controller/signup_test.go:229]: This function is not used anywhere, is it?

File: pkg/verification/verification_service.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+// ServiceImpl represents the implementation of the verification service.
+type ServiceImpl struct {
+	config     ServiceConfiguration
+	HttpClient *http.Client
+}
+
+type VerificationServiceOption func(svc *ServiceImpl)
+
+// NewVerificationService creates a service object for performing user verification
+func NewVerificationService(cfg ServiceConfiguration, opts ...VerificationServiceOption) Service {
```
Comment (alexeykazakov) [inline pkg/verification/verification_service.go:41]: You are not initiating any default http client? Looks a bit error prone. Http client is not among the required params but it is actually required.

File: pkg/verification/verification_service.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+		config: cfg,
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+	return s
+}
+
+// SendVerification sends a verification message to the specified user
```
Comment (alexeykazakov) [inline pkg/verification/verification_service.go:52]: Let's also mention that the function will add the verification code annotation to the signup.

File: pkg/verification/verification_service_test.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+		fromNumber:      fromNumber,
+		messageTemplate: configuration.DefaultVerificationMessageTemplate,
+	}
+}
+
+func (s *TestVerificationServiceSuite) TestVerify() {
+	defer gock.Off()
+	gock.New("https://api.twilio.com").
+		Reply(http.StatusNoContent).
+		BodyString("")
```
Comment (alexeykazakov) [inline pkg/verification/verification_service_test.go:68]: Let's check that the mock service got the expected api call: correct phone number and the verification code (the code is available in the annotation).

File: pkg/verification/verification_service_test.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+				v1alpha1.UserSignupPhoneNumberLabelKey: "+1NUMBER",
+			},
+		},
+		Spec: v1alpha1.UserSignupSpec{
+			Username: "sbryzak@redhat.com",
+		},
+	}
+
+	err := svc.SendVerification(ctx, userSignup)
+	require.Error(s.T(), err)
```
Comment (alexeykazakov) [inline pkg/verification/verification_service_test.go:125]: Check the error message too.

Comment (sbryzak) [reply pkg/verification/verification_service_test.go:125]: Done

File: pkg/verification/verification_service_test.go (hunk 6 above)
Comment (sbryzak) [reply pkg/verification/verification_service_test.go:68]: This was not easy, but I worked out a way to do it :)

File: pkg/verification/verification_service.go (hunk 5 above)
Comment (sbryzak) [reply pkg/verification/verification_service.go:52]: Done

File: pkg/verification/verification_service.go (hunk 4 above)
Comment (sbryzak) [reply pkg/verification/verification_service.go:41]: It's actually not required.  The very first thing that the `NewClient()` function does is create a default http client if it isn't set:

```
func NewClient(accountSid string, authToken string, httpClient *http.Client) *Client {
	if httpClient == nil {
		httpClient = defaultHttpClient
	}
```

File: pkg/controller/signup_test.go (hunk 3 above)
Comment (sbryzak) [reply pkg/controller/signup_test.go:229]: You're correct, it's not actually needed.  I've removed it.

File: pkg/verification/verification_service.go (hunk 4 above)
Comment (alexeykazakov) [reply pkg/verification/verification_service.go:41]: All good then!

File: pkg/verification/verification_service_test.go [hunk 8]
```diff
@@ -134,6 +157,7 @@ func (s *TestVerificationServiceSuite) createVerificationService() (verification
 		"CodeReady",
 	)
 
+	//	httpClient := &ClientMock{client: &http.Client{Transport: &http.Transport{}}}
```
Comment (alexeykazakov) [inline pkg/verification/verification_service_test.go:160]: I guess we don't need this comment?

Comment (sbryzak) [reply pkg/verification/verification_service_test.go]: Removed

**Pull request merged**